cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java

```diff
@@ -158,11 +169,21 @@ public void getOperationRun(
   @Path("/{id}/stop")
   public void failOperation(FullHttpRequest request, HttpResponder responder,
       @PathParam("namespace-id") String namespaceId,
-      @PathParam("id") String runId) {
-    // // TODO(samik, CDAP-20814) send the message to stop the operation
-    responder.sendString(HttpResponseStatus.OK,
-        String.format("Updated status for operation run %s in namespace '%s'.", runId,
-            namespaceId));
+      @PathParam("id") String runId)
+      throws OperationRunNotFoundException, IOException, BadRequestException, ForbiddenException {
+    checkSourceControlMultiAppFeatureFlag();
+    validateNamespaceId(namespaceId);
+    if (runId == null || runId.isEmpty()) {
+      throw new BadRequestException("runId cannot be empty");
+    }
+    // validate the run exists
+    OperationRunDetail runDetail =
```

Comment:
this sort of business logic should be in the lifecycle manager instead of in the handler. In general, http handlers should not be doing much more than deserializing user input and translating internal objects and errors into responses.

Reply:
Move the logic to lifecycle manager